src/lib/download-progress.ts

```diff
@@ -0,0 +1,88 @@
+import * as _ from 'lodash';
+import blink = require('./blink');
+import { DeviceStatus, TargetState } from '../types/state';
+import * as deviceState from '../device-state';
+import * as targetState from '../device-state/target-state';
+
+interface ServiceState {
+	id: string;
+	status: string;
+	download_progress: number | null;
+}
+
+const DOWNLOAD_BLINK_THROTTLE = 10000;
+const DOWNLOAD_BLINK_PERIOD_BASE = 100;
+const DOWNLOAD_BLINK_PERIOD_STEP = 10;
+
+const throttledBlink = _.throttle((progress: number) => {
+	blink.pattern.stop();
+	if (progress < 100) {
+		blink.pattern.start({
+			blinks: 1,
+			pause: 0,
+			onDuration:
+				DOWNLOAD_BLINK_PERIOD_BASE + DOWNLOAD_BLINK_PERIOD_STEP * progress,
+			offDuration:
+				DOWNLOAD_BLINK_PERIOD_BASE + DOWNLOAD_BLINK_PERIOD_STEP * progress,
+		});
+	}
+}, DOWNLOAD_BLINK_THROTTLE);
+
+export async function downloadProgress() {
+	const currentDeviceState = await deviceState.getStatus();
+
+	if (currentDeviceState.local?.update_pending) {
+		// Get target services and current state
+		const targetDeviceState = await targetState.get();
+		const targetServicesIDs = getTargetServicesIds(targetDeviceState);
+		const currentServicesStates = getCurrentServicesStates(currentDeviceState);
+
+		// Calculate overall progress
+		let progress = 0;
+		for (const id of targetServicesIDs) {
+			const service = currentServicesStates.find((s) => s.id === id);
+			if (service) {
+				if (['Downloaded', 'Installing', 'Running'].includes(service.status)) {
+					// States after 'Downloading' report 'null' download_progress. Patch those to 100%
+					progress += 100;
+				} else {
+					progress += service.download_progress ?? 0;
+				}
+			}
+		}
+		progress /= targetServicesIDs.length;
```

Comment:
Looking at the code it seems pretty certain that `targetServicesIDs.length` will never be 0 if there's an application running. What happens if there aren't any apps/services running (Device is "Factory build"), and you trigger a config change?

Reply:
I suppose this thought is only relevant if you keep `deviceState.on('change', downloadProgress)` since 'change' events are also emitted on config changes. If you refactor so that downloadProgress is only called on application changes though, this shouldn't be an issue.

Reply:
Good catch. I guess you are right, if the device is on "Factory build" there are no services on the target state so length would be 0. So triggering a config change could end the universe by dividing by 0. I didn't consider this scenario! 

It could inadvertedly be solved at the moment if the config change event does not set `update_pending` to `true`, do you know if that's the case? In any case this seems like a more solid approach :P :

> If you refactor so that downloadProgress is only called on application changes though, this shouldn't be an issue.

How would you tell a config change from an app change? Just test the `config` and `apps` keys on `state.local`?

Reply:
> It could inadvertedly be solved at the moment if the config change event does not set update_pending to true, do you know if that's the case?

Looks like a config change definitely [sets update_pending to true](https://github.com/balena-os/balena-supervisor/blob/0cc6920602468b824b08fddaa365c45ba1ff27db/src/device-state.ts#L843-L845).

> How would you tell a config change from an app change? Just test the config and apps keys on state.local?

If image fetch is the only time when you'd want blinking progress, I wonder if you can add the downloadProgress as a callback in only the image fetching step, since that's the only one where you need to download something. Maybe in this [onProgress callback](https://github.com/balena-os/balena-supervisor/blob/fd06c06092b61d349e8b40f66f847065368810e3/src/compose/images.ts#L132-L139) for images.triggerFetch? You can even pass in an `onFinish` callback to stop the blink pattern per what Felipe mentioned below about the `downloadProgress(95)` as an extra safeguard if you'd like.

Reply:
cc @pipex Do you think this would be a good idea to prevent what you mentioned [below](https://github.com/balena-os/balena-supervisor/pull/1728#discussion_r661578710)?

Reply:
The problem with the `onProgress` callback is that is per image, while on this case we need the download progress for the application release. I think is fine if we move the `targetServiceIDs` calculation outside of the if and add a `targetServiceIDs.length > 0` to the condition

Reply:
That's a good point ^ ignore my suggestion then :)

Reply:
That sounds good!